### What kind of programming style would affect salary

This section will look into the factors of programming style (what kind of IDE they use, space type) would contribute to salary, in order to determine the contribution, we build machine learning model to predict salary.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.metrics import r2_score, mean_squared_error
from collections import defaultdict
import seaborn as sns
%matplotlib inline

df = pd.read_csv('./survey_results_public.csv')
df.head()

Respondent                                       Professional  \
0           1                                            Student   
1           2                                            Student   
2           3                             Professional developer   
3           4  Professional non-developer who sometimes write...   
4           5                             Professional developer   

                ProgramHobby         Country      University  \
0                  Yes, both   United States              No   
1                  Yes, both  United Kingdom  Yes, full-time   
2                  Yes, both  United Kingdom              No   
3                  Yes, both   United States              No   
4  Yes, I program as a hobby     Switzerland              No   

                         EmploymentStatus  \
0  Not employed, and not looking for work   
1                      Employed part-time   
2                      Employed full-time   
3                      Employed full-time   
4                      Employed full-time   

                                     FormalEducation  \
0                                   Secondary school   
1  Some college/university study without earning ...   
2                                  Bachelor's degree   
3                                    Doctoral degree   
4                                    Master's degree   

                                  MajorUndergrad  \
0                                            NaN   
1       Computer science or software engineering   
2       Computer science or software engineering   
3  A non-computer-focused engineering discipline   
4       Computer science or software engineering   

                                          HomeRemote  \
0                                                NaN   
1              More than half, but not all, the time   
2  Less than half the time, but at least one day ...   
3  Less than half the time, but at least one day ...   
4                                              Never   

                CompanySize       ...       StackOverflowMakeMoney Gender  \
0                       NaN       ...            Strongly disagree   Male   
1        20 to 99 employees       ...            Strongly disagree   Male   
2  10,000 or more employees       ...                     Disagree   Male   
3  10,000 or more employees       ...                     Disagree   Male   
4        10 to 19 employees       ...                          NaN    NaN   

  HighestEducationParents                          Race         SurveyLong  \
0             High school  White or of European descent  Strongly disagree   
1       A master's degree  White or of European descent     Somewhat agree   
2   A professional degree  White or of European descent     Somewhat agree   
3       A doctoral degree  White or of European descent              Agree   
4                     NaN                           NaN                NaN   

  QuestionsInteresting QuestionsConfusing InterestedAnswers    Salary  \
0       Strongly agree           Disagree    Strongly agree       NaN   
1       Somewhat agree           Disagree    Strongly agree       NaN   
2                Agree           Disagree             Agree  113750.0   
3                Agree     Somewhat agree    Strongly agree       NaN   
4                  NaN                NaN               NaN       NaN   

   ExpectedSalary  
0             NaN  
1         37500.0  
2             NaN  
3             NaN  
4             NaN  

[5 rows x 154 columns]

### Data understanding and preparation
We will firstly extract all the programming style related columns and create dummy columns as they are categorical columns. 

In [2]:
sal_rm = df.dropna(subset=['Salary'], axis=0)
sal_nona = df.dropna(subset=['Salary', 'TabsSpaces', 'WorkStart', 'HaveWorkedLanguage', 
                             'HaveWorkedFramework', 'HaveWorkedDatabase', 'HaveWorkedPlatform', 'IDE', 
                             'AuditoryEnvironment', 'Methodology', 'VersionControl', 'CheckInCode'], axis=0)

In [3]:
sal_rm

Respondent            Professional  \
2               3  Professional developer   
14             15  Professional developer   
17             18  Professional developer   
18             19  Professional developer   
22             23  Professional developer   
25             26  Professional developer   
34             35  Professional developer   
36             37  Professional developer   
37             38  Professional developer   
52             53  Professional developer   
54             55  Professional developer   
57             58  Professional developer   
61             62  Professional developer   
64             65  Professional developer   
65             66  Professional developer   
69             70  Professional developer   
70             71  Professional developer   
72             73  Professional developer   
75             76  Professional developer   
76             77  Professional developer   
81             82  Professional developer   
83             84  Professional developer   
89             90  Professional developer   
90             91  Professional developer   
95             96  Professional developer   
97             98  Professional developer   
99            100  Professional developer   
102           103  Professional developer   
103           104  Professional developer   
104           105  Professional developer   
...           ...                     ...   
18945       18946  Professional developer   
18946       18947  Professional developer   
18947       18948  Professional developer   
18950       18951  Professional developer   
18955       18956  Professional developer   
18964       18965  Professional developer   
18984       18985  Professional developer   
18985       18986  Professional developer   
18989       18990  Professional developer   
18990       18991  Professional developer   
18996       18997  Professional developer   
19004       19005  Professional developer   
19005       19006  Professional developer   
19006       19007  Professional developer   
19011       19012  Professional developer   
19014       19015  Professional developer   
19016       19017  Professional developer   
19018       19019  Professional developer   
19020       19021  Professional developer   
19022       19023  Professional developer   
19039       19040  Professional developer   
19045       19046  Professional developer   
19046       19047  Professional developer   
19055       19056  Professional developer   
19073       19074  Professional developer   
19079       19080  Professional developer   
19086       19087  Professional developer   
19088       19089  Professional developer   
19089       19090  Professional developer   
19100       19101  Professional developer   

                                    ProgramHobby         Country  \
2                                      Yes, both  United Kingdom   
14                     Yes, I program as a hobby  United Kingdom   
17                                     Yes, both   United States   
18                     Yes, I program as a hobby   United States   
22                                            No          Israel   
25                     Yes, I program as a hobby   United States   
34                     Yes, I program as a hobby         Croatia   
36                     Yes, I program as a hobby       Argentina   
37                                     Yes, both         Germany   
52                     Yes, I program as a hobby          Brazil   
54                                            No         Germany   
57                                     Yes, both           Italy   
61                     Yes, I program as a hobby           India   
64                                            No     Netherlands   
65                     Yes, I program as a hobby          Poland   
69                     Yes, I program as a hobby           Spain   
70                     Yes, I program as a hobby   United St

In [4]:
sal_nona

Respondent            Professional  \
37             38  Professional developer   
52             53  Professional developer   
64             65  Professional developer   
72             73  Professional developer   
83             84  Professional developer   
133           134  Professional developer   
143           144  Professional developer   
146           147  Professional developer   
156           157  Professional developer   
173           174  Professional developer   
177           178  Professional developer   
180           181  Professional developer   
206           207  Professional developer   
211           212  Professional developer   
215           216  Professional developer   
234           235  Professional developer   
258           259  Professional developer   
260           261  Professional developer   
264           265  Professional developer   
284           285  Professional developer   
329           330  Professional developer   
331           332  Professional developer   
340           341  Professional developer   
349           350  Professional developer   
364           365  Professional developer   
373           374  Professional developer   
374           375  Professional developer   
378           379  Professional developer   
397           398  Professional developer   
398           399  Professional developer   
...           ...                     ...   
18630       18631  Professional developer   
18633       18634  Professional developer   
18655       18656  Professional developer   
18666       18667  Professional developer   
18668       18669  Professional developer   
18690       18691  Professional developer   
18696       18697  Professional developer   
18699       18700  Professional developer   
18718       18719  Professional developer   
18723       18724  Professional developer   
18756       18757  Professional developer   
18762       18763  Professional developer   
18773       18774  Professional developer   
18801       18802  Professional developer   
18804       18805  Professional developer   
18820       18821  Professional developer   
18821       18822  Professional developer   
18898       18899  Professional developer   
18902       18903  Professional developer   
18905       18906  Professional developer   
18909       18910  Professional developer   
18917       18918  Professional developer   
18940       18941  Professional developer   
18985       18986  Professional developer   
18996       18997  Professional developer   
19006       19007  Professional developer   
19011       19012  Professional developer   
19022       19023  Professional developer   
19079       19080  Professional developer   
19089       19090  Professional developer   

                                    ProgramHobby             Country  \
37                                     Yes, both             Germany   
52                     Yes, I program as a hobby              Brazil   
64                                            No         Netherlands   
72                                     Yes, both       United States   
83                                     Yes, both       United States   
133                                    Yes, both               India   
143                                    Yes, both      United Kingdom   
146                    Yes, I program as a hobby             Denmark   
156                                           No       United States   
173                                    Yes, both             Ukraine   
177                    Yes, I program as a hobby       United States   
180                                    Yes, both  Russian Federation   
206                                           No              Canada   
211                    Yes, I program as a hobby              Poland   
215                                    Yes, both               India   
234                    Yes, I program as a hobby       United St

We can see that if we remove all the null values, there would only be 1429 rows left comparing to the salary remove only option (5009 rows). Hence it's better to inpute the null values of categorical data with an additional N/A column.

In [5]:

dev_style_cols = ['TabsSpaces', 'WorkStart', 'HaveWorkedLanguage', 'HaveWorkedFramework', 'HaveWorkedDatabase',
            'HaveWorkedPlatform', 'IDE', 'AuditoryEnvironment', 'Methodology', 'VersionControl', 'CheckInCode']
prog_style_df = sal_rm[dev_style_cols]
salary = sal_rm['Salary']

def create_dummy_df(df, cat_cols, dummy_na):
    '''
    INPUT:
    df - pandas dataframe with categorical variables you want to dummy
    cat_cols - list of strings that are associated with names of the categorical columns
    dummy_na - Bool holding whether you want to dummy NA vals of categorical columns or not
    
    OUTPUT:
    df - a new dataframe that has the following characteristics:
            1. contains all columns that were not specified as categorical
            2. removes all the original columns in cat_cols
            3. dummy columns for each of the categorical columns in cat_cols
            4. if dummy_na is True - it also contains dummy columns for the NaN values
            5. Use a prefix of the column name with an underscore (_) for separating 
    '''
    for col in  cat_cols:
        try:
            # for each cat add dummy var, drop original column
            df = pd.concat([df.drop(col, axis=1), pd.get_dummies(df[col], prefix=col, prefix_sep='_', drop_first=True, dummy_na=dummy_na)], axis=1)
        except:
            continue
    return df

def CategorizeSalary(salary):
    result = pd.qcut(salary, 10, labels=False)
    return result

prog_style_dummy = create_dummy_df(prog_style_df, dev_style_cols, True)
prog_style_dummy.head()

TabsSpaces_Spaces  TabsSpaces_Tabs  TabsSpaces_nan  WorkStart_10:00 PM  \
2                   1                0               0                   0   
14                  1                0               0                   0   
17                  1                0               0                   0   
18                  0                0               1                   0   
22                  0                0               0                   0   

    WorkStart_11:00 AM  WorkStart_11:00 PM  WorkStart_1:00 AM  \
2                    0                   0                  0   
14                   0                   0                  0   
17                   0                   0                  0   
18                   0                   0                  0   
22                   0                   0                  0   

    WorkStart_1:00 PM  WorkStart_2:00 AM  WorkStart_2:00 PM       ...         \
2                   0                  0                  0       ...          
14                  0                  0                  0       ...          
17                  0                  0                  0       ...          
18                  0                  0                  0       ...          
22                  0                  0                  0       ...          

    VersionControl_Team Foundation Server  VersionControl_Visual Source Safe  \
2                                       0                                  0   
14                                      0                                  0   
17                                      0                                  0   
18                                      0                                  0   
22                                      0                                  0   

    VersionControl_Zip file back-ups  VersionControl_nan  \
2                                  0                   0   
14                                 0                   0   
17                                 0                   1   
18                                 0                   1   
22                                 0                   0   

    CheckInCode_A few times a week  \
2                                0   
14                               0   
17                               0   
18                               0   
22                               0   

    CheckInCode_Just a few times over the year  \
2                                            0   
14                                           0   
17                                           0   
18                                           0   
22                                           0   

    CheckInCode_Multiple times a day  CheckInCode_Never  \
2                                  1                  0   
14                                 1                  0   
17                                 0                  0   
18                                 0                  0   
22                                 1                  0   

    CheckInCode_Once a day  CheckInCode_nan  
2                        0                0  
14                       0                0  
17                       0                1  
18                       0                1  
22                       0                0  

[5 rows x 4051 columns]

### Modelling
We will firstly try to use linear regression model to predict salary. If linear regression does not work very well we will try logistic regression.

In [6]:
X = prog_style_dummy
min_max_scaler = preprocessing.MinMaxScaler()
#y = min_max_scaler.fit_transform(salary.values.reshape(-1, 1))
y_cat = CategorizeSalary(salary)

X_train, X_test, y_train, y_test = train_test_split(X, y_cat , test_size=.30, random_state=42)

lm_model = LinearRegression(normalize = True) # Instantiate
lm_model.fit(X_train, y_train) #Fit

tree_model = DecisionTreeRegressor()
tree_model.fit(X_train, y_train)

#Predict using your model
y_test_preds = lm_model.predict(X_test)
y_train_preds = lm_model.predict(X_train)

y_test_preds_tree = tree_model.predict(X_test)
y_train_preds_tree = tree_model.predict(X_train)

#Score using your model
test_score = r2_score(y_test, y_test_preds)
train_score = r2_score(y_train, y_train_preds)

test_score_tree = r2_score(y_test, y_test_preds_tree)
train_score_tree = r2_score(y_train, y_train_preds_tree)
print("The rsquared on the training data was {}.  The rsquared on the test data was {}.".format(train_score, test_score))
print("LOGIS: The rsquared on the training data was {}.  The rsquared on the test data was {}.".format(train_score_tree, test_score_tree))

The rsquared on the training data was 0.24269874093991195.  The rsquared on the test data was -1.8520610214013103e+29.
LOGIS: The rsquared on the training data was 0.9646497470281983.  The rsquared on the test data was -0.6903688041693936.


### Evaluation
From the training result, it seems that the linear regression model is performing quite bad. With logistic regression, the r2 score on training data is 0.965, which is quite good. However, the testing data shows the weakness of both the model. It means that the model overfitted on training data, and cannot be extended on new data.